contracts/starknet/authenticator/eth_tx.cairo

```diff
@@ -0,0 +1,78 @@
+%lang starknet
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.starknet.common.syscalls import call_contract
+from starkware.cairo.common.registers import get_fp_and_pc
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.memcpy import memcpy
+from starkware.cairo.common.hash_state import hash_init, hash_update
+from starkware.cairo.common.math import assert_not_equal
+from contracts.starknet.lib.eth_address import EthAddress
+
+# Address of the StarkNet Commit L1 contract which acts as the origin address of the messages sent to this contract.
+@storage_var
+func starknet_commit_address_store() -> (res : felt):
+end
+
+# Mapping between a commit and the L1 address of the sender.
+@storage_var
+func commit_store(hash : felt) -> (address : EthAddress):
+end
+
+@constructor
+func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    starknet_commit_address : felt
+):
+    starknet_commit_address_store.write(value=starknet_commit_address)
+    return ()
+end
+
+# Receives hash from StarkNet commit contract and stores it in state.
+@l1_handler
+func commit{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
+    from_address : felt, sender : EthAddress, hash : felt
+):
+    # Check L1 message origin is equal to the StarkNet commit address.
+    let (origin) = starknet_commit_address_store.read()
+    with_attr error_message("Invalid message origin address"):
+        assert from_address = origin
+    end
+    # Note: If the same hash is committed twice by the same sender, then the mapping will be overwritten but with the same value as before.
+    commit_store.write(hash, sender)
+    return ()
+end
+
+@external
+func execute{syscall_ptr : felt*, range_check_ptr, pedersen_ptr : HashBuiltin*}(
+    target : felt, function_selector : felt, calldata_len : felt, calldata : felt*
+):
+    alloc_locals
+    let (input_array : felt*) = alloc()
+    assert input_array[0] = target
+    assert input_array[1] = function_selector
+    memcpy(input_array + 2, calldata, calldata_len)
+    # Appending the length of the array to itself as the offchain version of the hash works this way
+    assert input_array[calldata_len + 2] = calldata_len + 2
+    let (hash_state_ptr) = hash_init()
+    let (hash_state_ptr) = hash_update{hash_ptr=pedersen_ptr}(
+        hash_state_ptr, input_array, calldata_len + 3
+    )
```

Comment:
I think we could share this code: https://github.com/snapshot-labs/sx-core/blob/0f81eddad82008762e4e6faf0205f337fdb0481f/contracts/starknet/space/space.cairo#L107-L118 by moving it to `lib/hash_pedersen` and call it here? wdyt?

Reply:
yes would make sense. although, i think we should add the length to the array inside the contract rather than  offchain so that the interface of `execute` is accurate. As with the length `calldata` is actually more like `calldata_plus_length`. The step savings of doing it offchain is pretty negligible.

Reply:
👍 go for it!